### ISSUE TITLE
Agent stuck in tiled navmesh

### ISSUE THREAD

**Issue by yelouafi** [isaac-mason/recast-navigation-js#453]
Using the following model with tile size of 32

[scene (11).glb.zip](https://github.com/user-attachments/files/18151322/scene.11.glb.zip)

the agent get stuck at some point on the stairs

https://github.com/user-attachments/assets/657a9a08-f212-46d6-a7ac-7e83f6550c85

on C++ demo the agent is fine. 

Did some debugging and seems the tile gen creates superficial connections between tiles on the stair and tiles below

![Screenshot 2024-12-16 at 12 16 13](https://github.com/user-attachments/assets/4c45ea3a-3ce5-4f19-aa89-9fdd4a17521c)

here the red poly is the clicked one and the green ones are the ones connected to it

on the C++ demo I used `find local neighbourhood` tool to show the connected polys

<img width="898" alt="Screenshot 2024-12-16 at 12 51 43" src="https://github.com/user-attachments/assets/3c90e3dc-b7bf-4231-859e-17c9aa512462" />

so looks like the issue is on the js side

**Comment by yelouafi** [comment]
looking at the detour cpp source seems tile connections are [computed here](https://github.com/recastnavigation/recastnavigation/blob/455a019e7aef99354ac3020f04c1fe3541aa4d19/Detour/Source/DetourNavMesh.cpp#L335-L338)

```cpp
if (!overlapSlabs(amin,amax, bmin,bmax, 0.01f, tile->header->walkableClimb)) continue;
```			

[in this lines in `overlapSlabs`](https://github.com/recastnavigation/recastnavigation/blob/455a019e7aef99354ac3020f04c1fe3541aa4d19/Detour/Source/DetourNavMesh.cpp#L59-L64)

```cpp
// Check for overlap at endpoints.
const float thr = dtSqr(py*2);
if (dmin*dmin <= thr || dmax*dmax <= thr)
return true;

return false;
```

seems to check if the other tile is vertically closer than `walkableClimb * 2`.  Not sure why it's checking the double of the value, but seems the issue is linked to `walkableClimb`.

**Comment by yelouafi** [comment]
In C++ demo walkable climb of the tile is rendered using the red square

<img width="898" alt="Screenshot 2024-12-16 at 12 51 43" src="https://github.com/user-attachments/assets/c8b9bb51-6277-4636-8f53-656d7f03f614" />

JS side I'm drawing the clciked tile's walkable climb using a scaled cylinder

![Screenshot 2024-12-16 at 16 20 01](https://github.com/user-attachments/assets/1ac416b6-ff40-4230-a956-ead07fedc7f1)

at this height * 2; it might explain why tiles are connected on js side

**Comment by yelouafi** [comment]
Found the issue!

It seems there's a bug in the generatros code, it's confusing the unit of mesure for when we bake the navmesh with for when we create the detour navmesh data. 

When we rasterize: height, climb ... are in voxel units. 
cf https://recastnav.com/structrcConfig.html#a9b6bbabb4857e3f4bf4428af86ee7673

When we create navigation data they must be in world units . 
cf https://recastnav.com/structdtNavMeshCreateParams.html#a2ccf071ca70fe5566ddd2f214bac7779
 
the generators code is using voxel units for both. In here

https://github.com/isaac-mason/recast-navigation-js/blob/b5679037b877003ec2c59332d23df4d5f4e53a7e/packages/recast-navigation-generators/src/generators/generate-solo-nav-mesh.ts#L392

it's just passing the recast value (in voxel units) to detour side which is expecting a wold units value.
with a voxel unit of 0.2 for example
A 1m climb height becomes 5 voxels
so we're actually creating a navmesh with 5m climb height

I think the confusion might've come from how units are handled in c++ demo vs in generators. In c++ demo GUI values for height, radius ... are in world units, and are only converted when passed to recast side (https://github.com/recastnavigation/recastnavigation/blob/455a019e7aef99354ac3020f04c1fe3541aa4d19/RecastDemo/Source/Sample_SoloMesh.cpp#L397), 

but otherwise passed as is to `dtNavMeshCreateParams` (https://github.com/recastnavigation/recastnavigation/blob/455a019e7aef99354ac3020f04c1fe3541aa4d19/RecastDemo/Source/Sample_SoloMesh.cpp#L702)

cc @isaac-mason

**Comment by isaac-mason** [comment]
Thank you for looking into this @yelouafi! Your description of the issue and proposed fix in the PR look sound to me, I'll get this merged and released shortly 🙂

**Comment by isaac-mason** [comment]
Seems this issue may also be resolved by this change: https://github.com/isaac-mason/recast-navigation-js/issues/381

**Comment by isaac-mason** [comment]
v0.37.0 has been released with the fix. Thanks again!

**Comment by yelouafi** [comment]
nice!
thanks for the rapid release

**Comment by isaac-mason** [comment]
No problems! 🙂

**Comment by niko-dellic** [comment]
@isaac-mason @yelouafi 

thanks for this- I've actually encountered a change in the off-mesh connection behavior in 0.37, which I suspect might come from the unit change:

When multiple off-mesh connections are stacked on top of each other (think elevator shaft), the lowest off mesh connection does not work

See below:

### 0.36.0
all of the bottom off mesh connections are working: here is a debug of the offmeshconnection flag:
![image](https://github.com/user-attachments/assets/2806f70c-cd02-4a54-8773-2ee861f9cfa7)

### 0.37.0
All of the bottom off mesh connection are not included. here is a debug of the offmeshconnection flags:
![image](https://github.com/user-attachments/assets/658c5253-6fe0-48ff-a750-686ddb17191d)


In this example, the first off mesh connection with the same xz coordiantes but the minumum y coordinate will not work. Upon inspecting the off mesh connections array, the connections are still being passed to the tiled generator (which is being performed on a web worker). 

offmeshConnections = [
{
startPosition: _Vector3 {x: 716.2453002929688, y: 146.875, z: -250.85630798339844}
endPosition: _Vector3 {x: 716.2453002929688, y: 190.625, z: -250.85630798339844}
bidirectional: true
},
{
startPosition: _Vector3 {x: 716.2453002929688, y: 190.625, z: -250.85630798339844}
endPosition: _Vector3 {x: 716.2453002929688, y: 234.375, z: -250.85630798339844}
bidirectional: true
},
{
startPosition: _Vector3 {x: 716.2453002929688, y: 234.375, z: -250.85630798339844}
endPosition: _Vector3 {x: 716.2453002929688, y: 278.125, z: -250.85630798339844}
bidirectional: true
},
....
]


Here are my recast settings, if needed.
cs | 0.32
maxAgentRadius : 0.15
walkableRadius : 1
walkableHeight : 2.25
walkableClimb : 1.5625
walkableSlopeAngle : 60
tileSize | 32
ch | 0.32
deleteGeometry : true
prune | true
numOffMeshConnections | 36


Reverting back to 0.36.0 does indeed result in the expected behavior. Any advice on the matter would be greatly appreciated, thanks!

**Comment by isaac-mason** [comment]
Hi @niko-dellic, I can find some time to look into this. If you're willing to share a reproduction of your issue that would help with investigation.

Offmesh links currently have a limtation that they can only connect to a neighbour tile. Maybe the unit change made in v0.37.0 has triggered this limitation?

**Comment by niko-dellic** [comment]
Hi @isaac-mason 

[index-debug.zip](https://github.com/user-attachments/files/18224504/index-debug.zip)

attached is a small glb of the scene and the offmesh connections. I have managed to further isolate the issue from further debugging:

cell size (cs) is the only parameter that affects this issue. In the attached model, a cell size of 0.26 fails to produce the bottom offmeshconnections. 0.26 does:

Failure (0.26):
![image](https://github.com/user-attachments/assets/bd7abcef-51f3-4d2f-8e88-fd8e3c1be371)


Success (0.25):
![image](https://github.com/user-attachments/assets/dc04c991-482e-4ee8-8732-0c1e08e6fd6a)

I've tried changing cell height and tilesize, but they have no impact on the off mesh connections. Here is all my config for reference:


cs | 0.26
maxAgentRadius | 0.15
walkableRadius | 1
walkableHeight | 2.25
walkableClimb | 1.923076923076923
walkableSlopeAngle | 60
tileSize | 26
ch | 0.26
numOffMeshConnections | 7

Let me know if there's anything else that I can help with. Thanks!

**Comment by niko-dellic** [comment]
Hey @isaac-mason, any thoughts? Thanks!

**Comment by isaac-mason** [comment]
Hey @niko-dellic, I've been out over the holiday period, I'll find some time in the new year to look at this, sorry :)

**Comment by niko-dellic** [comment]
@isaac-mason no problem, happy new year!

**Comment by niko-dellic** [comment]
Hey @isaac-mason, just bumping this in case you have some time to take a look. Thanks!

**Comment by isaac-mason** [comment]
Hi @niko-dellic,

I haven't looked into this deeply but can share this -

It seems the behaviour change is indeed related to the units fix for configs like walkableClimb.

Also worth mentioning that my general experience with off mesh connections is that they often need re-authoring when changing navmesh generation config.

I tried creating similar off mesh links using points returned from navMeshQuery.findClosestPoint and had no issues with that.

https://github.com/isaac-mason/recastnavigation/blob/599fd0f023181c0a484df2a18cf1d75a3553852e/Detour/Source/DetourNavMesh.cpp#L1024-L1050

```
{
    // old:
    // startPosition: {
    //   x: 56.09175491333008,
    //   y: 6.1955952644348145,
    //   z: -106.0022201538086,
    // },
    // new, with findClosestPoint
    startPosition: {
      x: 57.20161056518555,
      y: 6.3999924659729,
      z: -105.63432312011719,
    },
    // old:
    // endPosition: {
    //   x: 56.09175491333008,
    //   y: 49.945594787597656,
    //   z: -106.0022201538086,
    // },
    // new, with findClosestPoint
    endPosition: {
      x: 56.09175491333008,
      y: 50.239990234375,
      z: -106.0022201538086,
    },
    radius: 4.13,
    bidirectional: true,
    flags: 1,
    area: 1,
  },
```

<img width="691" alt="Image" src="https://github.com/user-attachments/assets/73740d78-0984-4c68-ac35-2eac790afdbb" />

https://codesandbox.io/p/live/b847379f-3fba-49ba-baa1-eb3c839a6591

looking more closely at dtNavMesh::connectExtOffMeshLinks, it internally calls findNearestPolyInTile with the provided positions, and uses these half extents for the query: 

```
const float halfExtents[3] = { targetCon->rad, target->header->walkableClimb, targetCon->rad };
```

https://github.com/isaac-mason/recastnavigation/blob/599fd0f023181c0a484df2a18cf1d75a3553852e/Detour/Source/DetourNavMesh.cpp#L473

I tried using the existing off mesh links you provided with an increased walkableClimb and off mesh links were created successfully:

<img width="501" alt="Image" src="https://github.com/user-attachments/assets/994e998c-66f2-4d88-a972-31dfd04404e4" />

**Comment by niko-dellic** [comment]
Hi @isaac-mason,

Thanks for taking a look into this, I finally got a chance to revisit this myself. I found that increasing the walkableClimb solved my issue, which makes sense given the changes included in 0.37 changed the y scaling. 

Perhaps you have some advice for generation config settings. I am making my navmesh for a crowd of agents that are average human height:

- If I set the walkableClimb to 3m, the offmeshconnections work, but this also implies that an agent can step up 3m. Ideally the walkable height would be 0.2m (stair height), but sloped surfaces then become jagged: settings: 

[cs: 0.3, ch:0.15, walkableClimb: 0.2]
![Image](https://github.com/user-attachments/assets/1e4cc6fe-ff84-4317-916d-c43aee808cd9)

if i set the walkableClimb to 3, I do not get the same issue on non-planar surfaces:  

[cs: 0.3, ch:0.15, walkableClimb: 3]
![Image](https://github.com/user-attachments/assets/01059b65-28c9-4a16-96ea-9df09a8b5ab7)

This also sets the walkableClimb to be greater than my current walkableHeight (which i've set to 2.6m to reflect the minimum height for most indoor rooms). I haven't been able to detect and errors that would result from this, but I can imagine this could cause agents to jump between floors if the height is greater than the walkableHeight but less than the walkableClimb.

Any advice on the matter? Thanks again!

**Comment by isaac-mason** [comment]
Hey @niko-dellic, I can suggest using the `DebugDrawer` utility to visualise the voxels in the heightfield nav mesh generation intermediate to help pick appropriate cell sizes/heights and agent properties. See: https://recast-navigation-js.isaacmason.com/?path=/story/debug-three-debug-drawer--heightfield-walkable

> If I set the walkableClimb to 3m, the offmeshconnections work, but this also implies that an agent can step up 3m. Ideally the walkable height would be 0.2m (stair height), but sloped surfaces then become jagged: settings:

As for this, all I can suggest right now is authoring off mesh connections again with points from findClosestPoint on your regenerated navmesh. It may be that the off mesh connections you are using are not exactly aligned with the latest generated navmesh and setting walkable height to be higher is just working around that.
